mikrotik/resource_bgp_instance.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceBgpInstance() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceBgpInstanceCreate,
+		Read:   resourceBgpInstanceRead,
+		Update: resourceBgpInstanceUpdate,
+		Delete: resourceBgpInstanceDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"as": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"client_to_client_reflection": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"comment": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"confederation_peers": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"disabled": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"ignore_as_path_len": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"out_filter": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			"redistribute_connected": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"redistribute_ospf": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"redistribute_other_bgp": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"redistribute_rip": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"redistribute_static": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"router_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"routing_table": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			"cluster_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"confederation": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceBgpInstanceCreate(d *schema.ResourceData, m interface{}) error {
+	instance := prepareBgpInstance(d)
+
+	c := m.(client.Mikrotik)
+
+	bgpInstance, err := c.AddBgpInstance(instance)
+	if err != nil {
+		return err
+	}
+
+	return bgpInstanceToData(bgpInstance, d)
+}
+
+func resourceBgpInstanceRead(d *schema.ResourceData, m interface{}) error {
+	c := m.(client.Mikrotik)
+
+	bgpInstance, err := c.FindBgpInstance(d.Id())
+
+	// TODO: Ignoring this error can cause all resources to think they
+	// need to be created. We should more appropriately handle this. The
+	// error where the record is not found is not actually an error and
+	// needs to be disambiguated from real failures
```

Comment:
We should never be copying this TODO. We need to properly handle this case.

We know that a `NotFound` error is returned in the case where the resource does not exist. It would look something like this.

```
	if _, ok := err.(client.NotFound); err != nil {
		d.SetId("")
		return nil
	}
```

Reply:
Do you want to perform a check on `ok`? Or is it enough knowing that `err` is something other than `nil` and then returning appropriately?

I ran into an issue trying to dereference `err`.